### DISCUSSION TITLE
Anyone else get this continuous build failure due to Generated/Metrics.swift ??

### DISCUSSION THREAD

**Discussion by gutley** [mozilla-mobile/firefox-ios#9511] [Q&A]
A few times a day, I fail to build the app with the following error:
```
error: input file '/Users/me/git/firefox-ios/Client/Generated/Metrics.swift' was modified during the build
error: input file '/Users/me/git/firefox-ios/Client/Generated/Metrics.swift' was modified during the build
Command CompileSwiftSources failed with a nonzero exit code
```
Once this happens, the only way to solve it is to clean the project and try again.

Its getting a bit irksome.

**Comment by lmarceau** [comment]
Hi @gutley,
Indeed it's an known issue we have that needs to be investigated. One workaround to avoid cleaning the project is to deleted the `Metrics.swift` file and build again. You can have an alias defined in your `.zshrc` or `.bashrc` for that like this:
```
# Fix the metrics issue by deleting the metrics
alias metrics="rm ~/your-own-path-to-the-project/firefox-ios/Client/Generated/Metrics.swift"
```
Hopefully we can resolve the issue soon! Thanks for asking about this.

**Comment by gutley** [comment]
Hi @lmarceau,

I've been doing some further digging and am beginning to understand what is happening without yet getting the underlying cause. I've also got a fairly brute force workaround which hasn't failed me yet in my testing.

**TL;DR** 

In the Glean SDK Generator Build phase, at some point after `rm -rf .venv; bash $SRCROOT/sdk_generator.sh -g MozillaAppServices` finishes, the Metrics File does indeed change, and I believe that there is something in the sdk_generator.sh that runs asynchronously to cause this. My current thinking is that for some reason the actual YAML parsing is done twice, generating two separate files at the same location. If you are lucky then sdk_generator.sh doesn't finish until the slower one does. If it finishes after the first file is done, then Xcode moves onto the Compiling Source phase, and at some point during this the other YAML finishes and writes out the Metrics file, causing Xcode to give the "was modified during the build" error.

**Evidence**

I modified the Xcode build phase as follows:

``` bash
rm -rf .venv; bash $SRCROOT/sdk_generator.sh -g MozillaAppServices

date
stat -x $SRCROOT/Client/Generated/Metrics.swift
sleep 30
stat -x $SRCROOT/Client/Generated/Metrics.swift
```

`date` is there to show when `sdk_generator.sh` says it has completed. The two `stat` calls help us see if anything fishy has happened to the file while the script sleeps for 30 seconds. If sdk_generator.sh was fully synchronous, then the output of the two calls should be mostly identical.

Here is an example of the output when the build succeeds:

```
✨ Your metrics are Glean! ✨
Wed Dec  1 12:51:49 GMT 2021
  File: "/Users/mgough/git/firefox-ios/Client/Generated/Metrics.swift"
  Size: 67239        FileType: Regular File
  Mode: (0644/-rw-r--r--)         Uid: (  502/  mgough)  Gid: (   20/   staff)
Device: 1,4   Inode: 53290153    Links: 1
Access: Wed Dec  1 12:51:49 2021
Modify: Wed Dec  1 12:51:49 2021
Change: Wed Dec  1 12:51:49 2021
(30 second sleep happens here)
  File: "/Users/mgough/git/firefox-ios/Client/Generated/Metrics.swift"
  Size: 67239        FileType: Regular File
  Mode: (0644/-rw-r--r--)         Uid: (  502/  mgough)  Gid: (   20/   staff)
Device: 1,4   Inode: 53290153    Links: 1
Access: Wed Dec  1 12:51:51 2021
Modify: Wed Dec  1 12:51:49 2021
Change: Wed Dec  1 12:51:49 2021
```

Here everything is as expected. The file has the same dates (ignoring AccessDate) and inode before and after the sleep and they are all <= the time when sdk_generator.sh finished.

Now lets look at an example where things went weird:
```
✨ Your metrics are Glean! ✨
Wed Dec  1 12:44:57 GMT 2021
  File: "/Users/mgough/git/firefox-ios/Client/Generated/Metrics.swift"
  Size: 67239        FileType: Regular File
  Mode: (0644/-rw-r--r--)         Uid: (  502/  mgough)  Gid: (   20/   staff)
Device: 1,4   Inode: 53267333    Links: 1
Access: Wed Dec  1 12:44:57 2021
Modify: Wed Dec  1 12:44:57 2021
Change: Wed Dec  1 12:44:57 2021
(30 second sleep happens here)
  File: "/Users/mgough/git/firefox-ios/Client/Generated/Metrics.swift"
  Size: 67239        FileType: Regular File
  Mode: (0644/-rw-r--r--)         Uid: (  502/  mgough)  Gid: (   20/   staff)
Device: 1,4   Inode: 53270351    Links: 1
Access: Wed Dec  1 12:45:25 2021
Modify: Wed Dec  1 12:45:24 2021
Change: Wed Dec  1 12:45:24 2021
```

Here we can see that not only is the file modification date of the Metrics file at the end of the script is after sdk_generator.sh supposedly finished, but we can also see (via the inode) that it is a completely different file to the one that was ready immediately after sdk_generator.sh. This is what leads me to believe that there is something async about sdk_generator.sh and that because it can result in two files being created, that the YAML is actually being done twice.

Unfortunately my knowledge of Python isn't going to help debug this any further. Hopefully someone else can see the 'obvious' error in the sdk_generator.sh script that causes this behaviour.

For now, in my fork, I am simply going to leave that 30 second sleep in as that ensures the the Metrics file is really complete before continuing on with the compiling. For most builds, Xcode dependency checking will not even call the Glean Script Phase so it doesn't impact overall build times unless the original .yaml file gets updated.

**Reply by lmarceau** [reply]
@gutley thanks a lot for having a look into the issue! I went to read a bit on this, and stumbled on a really interesting thread [here](https://github.com/mac-cain13/R.swift/issues/719) that looks to be pretty similar to the issue we have. 

Their conclusion is summarized in the comment [here](https://github.com/mac-cain13/R.swift/issues/719#issuecomment-937733804): 
> I think (hope) I’ve found the source of "modified during the build" errors; Xcode 13 now appears to run scripts during an Index 
> Build, while Xcode 12 didn't do that. This can cause R.swift to run again during Indexing, while a Build is simultaneously compiling 
> R.generated.swift.
>
> The way to prevent this is to add this check in the Run Script:
>
> ```
> if [ $ACTION != "indexbuild" ]; then
>  rswift generate my_dir/R.generated.swift
> fi
> ```

I tried it in our project and so far so good, but it would be great if someone else was testing it too before we bring this to others as a possible solution. In our build phase script, this translates to:

```
if [ $ACTION != "indexbuild" ]; then
    rm -rf .venv
    bash $SRCROOT/sdk_generator.sh -g MozillaAppServices
fi
```
The `R.swift` people also indicates in their [release notes](https://github.com/mac-cain13/R.swift/blob/master/Documentation/Migration.md#upgrading-to-60) that they:
> Uncheck "Based on dependency analysis" so that R.swift is run on each build

This results in the following build phase step:
![Screen Shot 2021-12-01 at 11 53 43 AM](https://user-images.githubusercontent.com/11338480/144279219-1f955123-ba98-44f1-be12-0c8913142f2a.png)

Let me know if you have time to try it, and if it's conclusive on your side.

**Comment by gutley** [comment]
@lmarceau Good find. That does seem the same problem as ours. I'll test it here tomorrow.

> Uncheck "Based on dependency analysis" so that R.swift is run on each build

Not sure that is needed in Firefox and will result in all builds being slower as that script takes a few seconds at the best of times. Since we specify input and output files, Xcode 'should' be happy with that.

I also saw that bit in the other thread from the Apple Dev about not using Build phases to generate code. Maybe a better solution would be if we could move that generation where its 'meant' to go.

**Reply by lmarceau** [reply]
@gutley Yeah worth checking for the `Based on dependency analysis` if it's necessary it or not. As for the build phases not being the right place to generate code, I noted the [WWDC video](https://developer.apple.com/videos/play/wwdc2021/10210/) and plan on having a listen at it soon 👍 I'm all up for improving our builds and configuring our project as Apple intends it to be used.

**Comment by gutley** [comment]
@lmarceau OK, I've run your new version here without issue so far, so looking good.

I did try moving things over to the build rules, but couldn't get the magic incantations to get that working, so for now I suggest we stick with the != indexbuild workaround.

I do suggest one small change to you script:

```sh
if [ $ACTION != "indexbuild" ]; then
    rm -rf .venv
    bash $SRCROOT/sdk_generator.sh -g MozillaAppServices
    echo "\n// Created via $ACTION action" >> "${SRCROOT}/Client/Generated/Metrics.swift"
fi
```

That echo appends a comment with the action type to the generated Metrics.swift file: e.g:

```swift
    .....
    enum Device {
        /// True if the device support device owner authentication
        /// with either biometrics or a passcode.
        static let authentication = BooleanMetricType( // generated from device.authentication
            category: "device",
            name: "authentication",
            sendInPings: ["metrics"],
            lifetime: .application,
            disabled: false
        )

    }

}

// Created via build action

```
So If this starts failing again, we'll be able to tell what action actually last created the file and if necessary exclude that action alongside indexbuild. (i.e in case Xcode introduce a new incompatable action type in future)

I also left the "Based on dependancy analysis" option turned on without issue, so suggest you do the same when you commit your changes.

**Reply by lmarceau** [reply]
Awesome, good idea for the echo! I'll chat with my teammates so we can do some more tests, and then hopefully open the PR soon. Thanks for the help on this!

**Comment by st3fan** [comment]
It sounds like the above fix works, but what about the two suggestions that the R.swift team got from Apple?

> Apple DTS made two suggestions:
>
>    1. Place a script in a Build pre-action in the schema
>    2. Consider using a custom build rule, as presented in this years WWDC: https://developer.apple.com/videos/play/wwdc2021/10210/?time=379

https://github.com/mac-cain13/R.swift/issues/719#issuecomment-936079241

Is that something we want to do? If this script is really running in the wrong place then maybe those suggestions are better?

**Reply by gutley** [reply]
@st3fan Whilst that would be the ideal long-term solution, I think this fix should be put in now anyway.

**Comment by lmarceau** [comment]
@gutley Update on this. 

The Glean team went forward to merge the indexbuild fix in the `generator_sdk.sh` file directly: https://bugzilla.mozilla.org/show_bug.cgi?id=1744504

So next time A-S is upgraded, we'll get that fix too. Long term plan is still to improve the way we generate builds as @st3fan suggested thought.
